### PR TITLE
Enhancement to test_rbdV2 and sdk layer for some commands

### PR DESCRIPTION
**Pull request by VasishtaShastry**
Added cluster selection option at test_config level and other minor enhancements

Particulars:
- Now cluster name can be mentioned at test_config level also.
With this enhancement now rbd mirroring test cases can be handled using single test_config file and need not to manage using multiple test_config files.
    modified:   tests/rbd/test_rbdV2.py

- Minor additions to SDK layer
    new file:   ceph/cephV2/ceph.py
    new file:   ceph/cephV2/osd.py
    new file:   ceph/cephV2/pool.py
    new file:   ceph/rbd/feature.py
    new file:   ceph/rbd/mirror/bootstrap.py
    new file:   ceph/rbd/mirror/peer.py
    new file:   ceph/rbd/mirror/pool.py

- Removed Utlis methods and added couple of frequently used sub-commands
    modified:   ceph/rbd/rbd.py
- Moved clone to rbd.py as its a sub command of rbd not snapshot
    modified:   ceph/rbd/snapshot.py
- removed layering.py as there are no subcommand named layering for rbd
    deleted:    ceph/rbd/layering.py
- added a temorary test_config to refer as an example
    new file:   tests/rbd_mirror/test_configs/temp_image_mirroring.yaml
Signed-off-by: Vasishta <vashastr@redhat.com>

# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by HaruChebrolu** [comment]
Please remove unused imports and add proper Google style documentation for the modules that you used

File: ceph/cephV2/osd.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+and modules for sub-commands of `ceph osd ` would use this module.
+"""
+
+from ceph.cephV2.ceph import Ceph
+from utility.log import Log
+
+log = Log(__name__)
+
+
+class Osd(Ceph):
```
Comment (HaruChebrolu) [inline ceph/cephV2/osd.py:13]: use composition inheritance

File: ceph/rbd/mirror/bootstrap.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+        Returns:
+            out: Output after execution of command
+                 Out would contain the key.
+            err: Error after execution of command
+        """
+        cmd = f'rbd mirror pool peer bootstrap create {args["pool_name"]}'
+        if args.get("site_name", None):
+            cmd = cmd + f" --site-name {args['site_name']}"
+        if args.get("file_path", None):
```
Comment (HaruChebrolu) [inline ceph/rbd/mirror/bootstrap.py:38]: IMO, we can port this write_to_file implementation to an utility which can be reusable. What do you say?

File: ceph/rbd/snapshot.py [hunk 3]
```diff
@@ -22,7 +22,7 @@ def create(self, args):
         pool_name = args["pool_name"]
         image_name = args["image_name"]
         cmd = f"rbd snap create {pool_name}/{image_name}@{snap_name}"
-        if self.exec_cmd(cmd=cmd):
+        if not self.exec_cmd(cmd=cmd):
```
Comment (HaruChebrolu) [inline ceph/rbd/snapshot.py:25]: Please revisit this.

File: utility/copy_file.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+class CopyFile:
+    @staticmethod
+    def run(kw, args):
+        """Copy the file from one instance to another.
+
+        Copy the specified file from source to destination instance.
+
+        Usage
+        - copy_file: True
```
Comment (HaruChebrolu) [inline utility/copy_file.py:14]: Please update documentation of usage

File: utility/copy_file.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+          source: ceph-rbd1
+          dest: ceph-rbd2
+          path: ./dummy_file
+        """
+
+        cluster = kw["ceph_cluster_dict"][args["source"]]
+
+        for node in cluster:
```
Comment (HaruChebrolu) [inline utility/copy_file.py:22]: IMO, we can have this in an utility

File: utility/copy_file.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+            if node.role == "client":
+                source_node = node
+                break
+
+        cmd = f"sleep 10;cat {args['path']}"
```
Comment (HaruChebrolu) [inline utility/copy_file.py:27]: Do we need this? If yes, I think we can handle this from workflow layer

**Comment by VasishtaShastry** [comment]
After conversations yesterday, main aspects to implement are -

- [ ]  Convert all classes from multiple inheritance to composite inheritance
- [ ] Revisit test_rbdV2.py and modify to accommodate composite inheritance

File: ceph/cephV2/osd.py (hunk 1 above)
Comment (VasishtaShastry) [reply ceph/cephV2/osd.py]: Have migrated all classes to composite inheritance
Please review

File: ceph/rbd/mirror/bootstrap.py (hunk 2 above)
Comment (VasishtaShastry) [reply ceph/rbd/mirror/bootstrap.py]: Yes, Moved it out of sdk layer and to utility
Sounds good to stick with only commands in sdk layer

File: ceph/rbd/snapshot.py (hunk 3 above)
Comment (VasishtaShastry) [reply ceph/rbd/snapshot.py]: Sorry, had missed it
Removed it now, It had come when we had created a blueprint in initial commit

File: utility/copy_file.py (hunk 4 above)
Comment (VasishtaShastry) [reply utility/copy_file.py]: Done.

File: utility/copy_file.py (hunk 5 above)
Comment (VasishtaShastry) [reply utility/copy_file.py]: changed it to calling an existing method from ceph object (will finalize after testing)

File: utility/copy_file.py (hunk 6 above)
Comment (VasishtaShastry) [reply utility/copy_file.py]: Yes, moved it out

**Comment by VasishtaShastry** [comment]
Currently I'm blocked with 
1) modifying sdk layer methods as I'm unsure of initialising cli class for calling exec_cmd
2) remodeling workflow layer handler for test_rbdv2 to accommodate composite inheritance

**Comment by psathyan** [comment]
@VasishtaShastry is there a high level design or spec that I can refer to help you to get unblocked?

With that being said, there is an ongoing parallel initiative being driven by @HaruChebrolu where in the decoupling of CI, test keywords, test workflows and test data. You could reach out to @HaruChebrolu for further information and see if that fits into your immediate requirement.

The discussion could result in further improvements to your initiative or the on-going.

**Comment by VasishtaShastry** [comment]
(Adding summary from offline conversation)
Work in this PR was started as an extender for PR 1323 in which a blueprint was provided for automating testcases of rbd FG in the model of (SDK+Workflow layer).
With multiple drawbacks found during the implementation of workflows layer (tests/rbd/test_rbdV2.py) and parallel initiatives that have been started as mentioned above and betterment in long run, it makes sense to follow the ongoing new initiatives which are based on updated design models for automation.

**Comment by VasishtaShastry** [comment]
Discussed at RBD FG level and decided to continue using previous approach till we move to a data driven model.

**Pull request closed without merging**